src/axom/quest/DistributedClosestPoint.hpp

```diff
@@ -680,207 +680,197 @@ class DistributedClosestPointImpl
       isBVHTreeInitialized(),
       "BVH tree must be initialized before calling 'computeClosestPoints");
 
-    BoxType queryPartitionBb = computeMeshBoundingBox(queryMesh, coordset);
+    BoxType myQueryBb = computeMeshBoundingBox(queryMesh, coordset);
+    BoxArray allQueryBbs;
+    gatherBoundingBoxes(myQueryBb, allQueryBbs);
 
     std::map<int, std::shared_ptr<conduit::Node>> xferNodes;
-    std::list<std::shared_ptr<conduit::Node>> skipNodes;
 
     // create conduit node containing data that has to xfer between ranks.
     // The node will be mostly empty if there are no query points on this rank
     {
       xferNodes[m_rank] = std::make_shared<conduit::Node>();
       conduit::Node& xferNode = *xferNodes[m_rank];
       copy_query_node_to_xfer_node(queryMesh, xferNode, coordset);
-      put_bounding_box_to_conduit_node(queryPartitionBb, xferNode["aabb"]);
+      put_bounding_box_to_conduit_node(myQueryBb, xferNode["aabb"]);
+      xferNode["homeRank"] = m_rank;
+    }
+
+    {
+      conduit::Node& xferNode = *xferNodes[m_rank];
+      computeLocalClosestPointsByPolicy(xferNode);
+    }
 
-      if(m_isVerbose)
+    const auto& myObjectBb = m_objectPartitionBbs[m_rank];
+    int remainingRecvs = 0;
+    for(int r = 0; r < m_nranks; ++r)
+    {
+      if(r != m_rank)
       {
-        internal::dump_node(xferNode,
-                            fmt::format("round_{}_r{}_begin.json", 0, m_rank));
+        const auto& otherQueryBb = allQueryBbs[r];
+        double sqDistance =
+          axom::primal::squared_distance(otherQueryBb, myObjectBb);
+        if(sqDistance <= m_sqDistanceThreshold)
+        {
+          ++remainingRecvs;
+        }
       }
     }
```

Comment:
Pre-computing the next recipient is a bit inefficient.  We'd have to assess N `BoundingBox` distances and store up to N ranks indices.  We have to communicate that O(N) data as well, if we don't want to re-compute it on each rank the query `Node` lands on.  Moreover, a rank never needs to know all the other ranks a query `Node` is close to.  It just needs the next rank.  The loop in `next_recipient` `m_nranks` long, but we expect it to break very early.

Reply:
I was thinking that if you sort it as a preprocess, you could avoid work later on by not sending data to ranks that provably cannot have the closest point.

I suppose there is a tradeoff between how many ranks (`N`) we expect to run on  and how many ranks (on average) we expect to send data to.  For degenerate configurations, we might need to send data to all ranks (e.g. a point near the center of a sphere), but for many configs, we can probably prune most of the ranks during this preprocessing step.

In any event, this can be refactored/optimized in the future, as necessary/desired.

Reply:
For the default infinite threshold, we don't gain anything by preprocessing, because there's no knowledge to be gained.  The `next_recipient` loop would return at once.
If the number allowed by the threshold << number of ranks, there's lots go gain.  However, If I read your suggestions right, it amounts to calling `next_recipient` ahead of time instead of as needed.  I think we'd end up doing the same amount of computations.
My main concern is clarity, since this refactoring is supposed to improve that.  Do you think the preprocessing would help with clarity?  I see that it only replaces calling `next_recipient` with a look-up, but maybe I'm missing something.